presto-main/src/main/java/com/facebook/presto/operator/scalar/MergingDigest.java

```diff
@@ -0,0 +1,799 @@
+    public Collection<Centroid> centroids()
+    {
+        compress();
+        return new AbstractCollection<Centroid>()
+        {
+            @Override
+            public Iterator<Centroid> iterator()
+            {
+                return new Iterator<Centroid>()
+                {
+                    int i;
+
+                    @Override
+                    public boolean hasNext()
+                    {
+                        return i < activeCentroids;
+                    }
+
+                    @Override
+                    public Centroid next()
+                    {
+                        Centroid rc = new Centroid(mean[i], (int) weight[i], data != null ? data.get(i) : null);
+                        i++;
+                        return rc;
+                    }
+
+                    @Override
+                    public void remove()
+                    {
+                        throw new UnsupportedOperationException("Default operation");
+                    }
+                };
+            }
+
+            @Override
+            public int size()
+            {
+                return activeCentroids;
+            }
+        };
+    }
+
+    @Override
+    public double compression()
+    {
+        return publicCompression;
+    }
+
+    @Override
+    public int byteSize()
+    {
+        compress();
+        // format code, type code, min (double), max (double), compression(double), #centroids(int),
+        // then two doubles per centroid
+        return activeCentroids * 16 + 30;
+    }
+
+    @SuppressWarnings("WeakerAccess")
+    public ScaleFunction getScaleFunction()
+    {
+        return scale;
+    }
+
+    public enum Encoding
+    {
+        VERBOSE_ENCODING((byte) 0), SMALL_ENCODING((byte) 1);
+
+        private final byte code;
+
+        Encoding(byte code)
+        {
+            this.code = code;
+        }
+    }
+
+    @Override
+    public Slice serialize()
+    {
+        SliceOutput sliceBuilder = new DynamicSliceOutput(byteSize());
+
+        compress();
+        sliceBuilder.writeByte(0); // version 0 of T-Digest serialization                     // 1
+        sliceBuilder.writeByte(0); // represents the underlying data type of the distribution // + 1
+        sliceBuilder.writeDouble(min);                                                              // + 8
+        sliceBuilder.writeDouble(max);                                                              // + 8
+        sliceBuilder.writeDouble(publicCompression);                                                // + 8
+        sliceBuilder.writeInt(activeCentroids);                                                     // + 4 = 30
+        for (int i = 0; i < activeCentroids; i++) {
+            sliceBuilder.writeDouble(weight[i]);
+            sliceBuilder.writeDouble(mean[i]);
+        }
+
+        return sliceBuilder.slice();
+    }
+
+    @SuppressWarnings("WeakerAccess")
+    public static MergingDigest deserialize(Slice slice)
```

Comment:
By convention we put our deserialization logic in a constructor which takes in a slice